src/server.rs

```diff
@@ -0,0 +1,475 @@
+    mod_Message::Block as ProtoBlock, mod_Message::Wantlist as ProtoWantlist, Message,
+};
+use crate::utils::stream_protocol;
+use crate::{Event, Result, StreamRequester, ToBehaviourEvent, ToHandlerEvent};
+
+type Sink = FramedWrite<libp2p_swarm::Stream, Codec>;
+type BlockWithCid<const S: usize> = (CidGeneric<S>, Vec<u8>);
+
+#[derive(Debug)]
+pub(crate) struct ServerBehaviour<const S: usize, B>
+where
+    B: Blockstore + Send + Send,
+{
+    protocol: StreamProtocol,
+    store: Arc<B>,
+
+    peers_waitlists: FnvHashMap<PeerId, PeerWantlist<S>>,
+    global_waitlist: FnvHashMap<CidGeneric<S>, SmallVec<[PeerId; 1]>>,
+
+    outgoing_queue: Vec<BlockWithCid<S>>,
+    outgoing_event_queue: VecDeque<ToSwarm<Event, ToHandlerEvent>>,
+
+    blockstore_tasks: FuturesUnordered<BoxFuture<'static, BlockstoreResult<S>>>,
+    blockstore_tasks_abort_handles: FnvHashMap<(PeerId, CidGeneric<S>), AbortHandle>,
+}
+
+#[derive(Debug)]
+enum BlockstoreResult<const S: usize> {
+    Get(
+        PeerId,
+        CidGeneric<S>,
+        Result<Option<Vec<u8>>, BlockstoreError>,
+    ),
+    Cancelled,
+}
+
+#[derive(Debug, Default)]
+struct PeerWantlist<const S: usize>(FnvHashSet<CidGeneric<S>>);
+
+impl<const S: usize> PeerWantlist<S> {
+    pub fn process_wantlist(&mut self, wantlist: ProtoWantlist) -> Vec<WishlistChange<S>> {
+        // XXX quietly drop invalid entries from wantlist, do we care about logging?
+        if wantlist.full {
+            let wanted_cids = wantlist
+                .entries
+                .into_iter()
+                .filter_map(|e| {
+                    if e.cancel {
+                        return None;
+                    }
+                    CidGeneric::try_from(e.block).ok()
+                })
+                .collect();
+
+            return self.wantlist_replace(wanted_cids);
+        }
+
+        let mut results = Vec::with_capacity(wantlist.entries.len());
+
+        for entry in wantlist.entries {
+            let Ok(cid) = CidGeneric::try_from(entry.block) else {
+                continue;
+            };
+
+            if !entry.cancel {
+                if self.0.insert(cid) {
+                    results.push(WishlistChange::WantCid(cid));
+                }
+            } else if self.0.remove(&cid) {
+                results.push(WishlistChange::DoesntWantCid(cid))
+            }
+        }
+
+        results
+    }
+
+    fn wantlist_replace(&mut self, cids: FnvHashSet<CidGeneric<S>>) -> Vec<WishlistChange<S>> {
+        let wishlist_delta = cids
+            .difference(&self.0)
+            .map(|cid| WishlistChange::WantCid(*cid))
+            .chain(
+                self.0
+                    .difference(&cids)
+                    .map(|cid| WishlistChange::DoesntWantCid(*cid)),
+            );
+
+        let wishlist_delta = wishlist_delta.collect();
+
+        self.0 = cids;
+
+        wishlist_delta
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum WishlistChange<const S: usize> {
```

Comment:
I found 3 different terms for that, wantlist, wishlist and waitlist. Could be unify them to wantlist?

Reply:
😅 
Merged wantlist and wishlist into `wantlist` since you're right, they're the same and it's confusing.
Sendlist on the other hand refers to messages we want to send back, I've replaced that with `pending_outgoing_messages` (or `pending_messages`/`messages`, in cases where context makes it obvious what we're talking about).